ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,483 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+    SIZE_T UpdatedLimit, FinalLimit, ReturnedQuotas;
 
-    /* Guard our quota in a spin lock */
-    KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
+    /* Sanity checks */
+    ASSERT(QuotaType < PsQuotaTypes);
+    ASSERT((SSIZE_T)Amount >= 0);
 
-    if (Process->QuotaUsage[QuotaType] + Amount >
-        Process->QuotaBlock->QuotaEntry[QuotaType].Limit)
+    /* Are we within the bounds of quota limit? */
+    if (QuotaBlock->QuotaEntry[QuotaType].Usage + Amount >
+        QuotaBlock->QuotaEntry[QuotaType].Limit &&
+        QuotaBlock != &PspDefaultQuotaBlock)
     {
-        DPRINT1("Quota exceeded, but ROS will let it slide...\n");
+        /* We aren't... Is this a page file quota charging? */
+        if (QuotaType == PsPageFile)
+        {
+            /* It is, return the appropriate status code */
+            DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Quota amount exceeds the limit on page file quota (limit -- %lu || amount -- %lu)\n",
+                    QuotaBlock->QuotaEntry[QuotaType].Limit, Amount);
+            return STATUS_PAGEFILE_QUOTA_EXCEEDED;
+        }
+
+        /* Guard ourselves quota in a spin lock */
+        KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
+
+        /*
+         * This is not a page file charge. What we can do at best
+         * in this scenario is to attempt to raise (expand) the
+         * quota limit charges of the block.
+         */
+        if (!MmRaisePoolQuota(QuotaType,
+                              QuotaBlock->QuotaEntry[QuotaType].Limit,
+                              &UpdatedLimit))
+        {
+            /*
+             * We can't? It could be that we must free
+             * up some resources in order to raise the
+             * limit, which in that case we must return
+             * the excess of quota that hasn't been
+             * returned. If we haven't returned anything
+             * then what we're doing here is futile.
+             * Bail out...
+             */
+            PspReturnExcessQuotas(QuotaType, &ReturnedQuotas);
+            if (ReturnedQuotas == 0)
+            {
+                DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Failed to free some resources in order to raise quota limits...\n");
+                KeReleaseSpinLock(&PspQuotaLock, OldIrql);
+                return STATUS_QUOTA_EXCEEDED;
+            }
+
+            /* Try to raise the quota limits again */
+            MmRaisePoolQuota(QuotaType,
+                             QuotaBlock->QuotaEntry[QuotaType].Limit,
+                             &UpdatedLimit);
+        }
+
+        /* Release the lock */
         KeReleaseSpinLock(&PspQuotaLock, OldIrql);
-        return STATUS_SUCCESS;
-        //return STATUS_QUOTA_EXCEEDED; /* caller raises the exception */
+
+        /* Enforce a new raised limit */
+        FinalLimit = QuotaBlock->QuotaEntry[QuotaType].Limit - UpdatedLimit;
+        InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
```

Comment:
This is inconsistent with the re-fetching of `Limit` done above. I think `Limit` needs to follow one of two rules (it seems to be the first on Windows`*`, but I think the second would work fine):

Either, `Limit` updates need to always happen under lock. In this case, this CompareExchange should move before the `KeReleaseSpinLock`, but is basically safe. It could then just be an InterlockedExchange or an InterlockedExchangeAdd, I believe.

Or, `Limit` updates can happen outside the lock as long as they are interlocked. In this case:
* The code above needs to use a fetch-once approach instead of re-reading `Limit`
* This CompareExchange needs to use the once-fetched value
* There needs to be a way to recover from CompareExchange failure scenario (e.g. with a loop; or simply by not using CompareExchange. You could calculate `AddedLimit = UpdatedLimit - OldLimit` here and just use an `Interlocked(Exchange)Add`)


`* I'm actually not sure -- they seem to both hold the lock everywhere they update Limit, and use safe interlocked operations, which means I may be missing some places where Limit is updated without the lock held`